Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,254 @@
+#include "Projectiles.h"
+#include "../Helpers.h"
+
+
+Projectile::Projectile(size_t id) : IMovable(id)
+{
+    SetPassable(D_ALL, Passable::AIR);
+    tickSpeed = 0.25;
+    damage = 0;
+    SetSprite("icons/projectiles.dmi");
+    v_level = 5;
+}
+void Projectile::Process()
+{
+    if(set_target_ == 0)
+    {
+       return;
+    }
+    int z = current_step_;
+    Dir step = VDirToDir(movement_[z]);
+
+   if(!TryMove(step))
+   {
+       this->Delete();
+       return;
+   }
+   current_step_++;
+   if(!(current_step_ < movement_.size()) && !reached_target )
+   {
+       current_step_ = 0;
+       VDir temp = movement_[0];
+       movement_.clear();
+       reached_target = true;
+       movement_.push_back(temp);
+   }
+   if(!(current_step_<movement_.size()) && reached_target )
+   {
+   current_step_ = 0;
+   }   
+}
+
+bool Projectile::CheckPassable()
+{
+    id_ptr_on<Projectile> p = this->GetId();
+    PassableLevel loc = GetPassable(dMove);
+    if (loc != Passable::FULL)
+    {
+        SetPassable(dMove, Passable::FULL);
+    }
+    if (!CanPass(owner->GetPassable(dMove), passable_level))
+    {
+        owner->AttackByP(p);
+        if (loc != Passable::FULL)
+        {
+            SetPassable(dMove, loc);
+        }
+        return false;
+    }
+    if (loc != Passable::FULL)
+    {
+        SetPassable(dMove, loc);
+    }
+
+    auto neighbour = owner->GetNeighbour(dMove);
+    if (   !CanPass(neighbour->GetPassable(D_ALL), passable_level)
+        || !CanPass(neighbour->GetPassable(helpers::revert_dir(dMove)), passable_level))
+    {
+        neighbour->AttackByP(p);
+        return false;
+    }
+    
+    return true;
+}
+void Projectile::Bump(id_ptr_on<IMovable> item)
+{
```

Comment:
Yes, the `Bump` function uses `IMovable` as argument. And overall logic suitable for projectiles - `Bump` is called when something (human, projectile or anything) hits object. So that's why I think that `AttackByP` is not really needed.